Development

matheusneder opened the pull request:
The Upcomming Movies App MVP.

Commit:
Initial commit

Commit:
Fixed .gitignore

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update CoreException.cs

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Update README.md

matheusneder committed:
Added CoreException messagens, adjusted some project configuration and reviewed intellisense texts.

matheusneder committed:
Merge branch 'development' of https://github.com/matheusneder/upcomming-movies into development

matheusneder committed:
Update README.md

Commit:
Added search feature.

Commit:
Fixed some mispelled words in coments.

Commit:
Added sreenshots

Commit:
Renamed Source to source

matheusneder committed:
Update ApplicationInstaller.cs

matheusneder committed:
Update CoreException.cs

matheusneder committed:
Update MainPageViewModel.cs

arctouch-eltonsantana commented on source/UpcommingMovies.Infra.TheMovieDb.Test/MovieDiscoverServiceTest.cs:
```diff
@@ -0,0 +1,45 @@
+        public async Task TestRetrieveUpCommingMovies()
+        {
+            var upcomming = await _movieDiscoverService.RetrieveMoviesAsync(1);
+            var x = upcomming.Movies.Select(m => m.BuildBackdropImageUri(500)).ToArray();
```
Where are the Asserts?

matheusneder replied:
I did this tests to support the movie db service client development. Since it depends of the movie db endpoint, it's not well suitable to unit test the app. I agree that is missing decoupled unit tests and its Asserts.

arctouch-eltonsantana commented on source/UpcommingMovies.Infra.TheMovieDb/Services/MovieDiscoverService.cs:
```diff
@@ -0,0 +1,247 @@
+
+            try
+            {
+                var jObject = JObject.Parse(jsonResult);
```
Why did you choose this approach instead of object deserialization?

matheusneder replied:
As a design choice, I used JObject in order to build my domain objects in a custom way. It also could be done by JsonConverter aproach (extending JsonCoverter in order to write a custom converter), but I choose the first aproach thinking about simplicity.

arctouch-eltonsantana commented on source/UpcommingMovies.UI/UpcommingMovies.UI/Models/MovieListItem.cs:
```diff
@@ -0,0 +1,40 @@
+            Movie = movie;
+            Genres = string.Join(", ", movie.Genres.Select(g => g.Name));
+            HasGenres = movie.Genres.Any();
+            Image = movie.BuildBackdropImageUri(150) ?? movie.BuildPosterImageUri(150);
```
The images take a long time to load because you are using the backdrop instead of the poster.

matheusneder replied:
The BuildBackdropImageUri and BuildPosterImageUri methods takes width as parameter. For the given width, it looks for nearest width available for image type (as supplied by configuration). I choose backdrop images as prefered image instead of poster by an arbitrary criteria, but now that you address this, I see that the minimum width for backdrop images is 300px, that is considerable larger than the 92px of the minimum width for posters. Sorry, my fault.

arctouch-eltonsantana reviewed (commented):
Overall, a pretty good implementation employing Xamarin.Forms, Prism and Unity.

Strong points:
- Infinite Scroll
- Search feature
- Globalization
- Code is pretty clean

Drawbacks
- There is no connection handling
- There is no wait indicator
- Images take a long time to load
- Images and Http requests not cached
- Using JObject instead of deserializing object
- There is no 'Asserts' on unit tests

matheusneder commented on source/UpcommingMovies.Infra.TheMovieDb/Services/TheMovieDbServiceBase.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <exception cref="RemoteServiceException">
+        /// Response gives any http status but 200 family, 401, 404 or 429.
+        /// </exception>
+        protected async Task<string> GetAsync(string relativePath, IDictionary<string, string> query = null, bool supressLanguage = false)
```
@arctouch-eltonsantana 
Connection exception handling is done here. The exception types thrown by this method extends CoreException which all will being cautch by the UI logic. The CoreException defines the FriendlyMessage property. FriendlyMessage property is overridden by derived classes in order to provide friendly and internationalized messages to the consumer. The UI layer show this messages to the user in a DisplayAlert dialog.

Commit:
Adjusted to show poster image (it's smaller) instead of backdrop image on movie list.

matheusneder commented on source/UpcommingMovies.UI/UpcommingMovies.UI/Models/MovieListItem.cs:
```diff
@@ -14,7 +14,7 @@ public MovieListItem(Movie movie)
             Genres = string.Join(", ", movie.Genres.Select(g => g.Name));
             HasGenres = movie.Genres.Any();
-            Image = movie.BuildBackdropImageUri(150) ?? movie.BuildPosterImageUri(150);
+            Image = movie.BuildPosterImageUri(90);
```
No more backdrop images on the list!

Commit:
Cache images!

matheusneder reviewed (commented):
@arctouch-eltonsantana 
Added image cache explicit configuration. At first, this is not necessary since ImageSource caches images localy for 24 hours by default. 

*"Caching is enabled by default and will store the image locally for 24 hours."* (https://developer.xamarin.com/guides/xamarin-forms/user-interface/images/#Downloaded_Image_Caching)

After you adressed this, I did some investigation to ensure that ImageSource really makes cache as describled at documentation. By using fiddler as device's proxy, I could look at the trafic and I have confirmed that the cache is been done as describled by the doc, even if not explicity configured.

Commit:
Added wait indicator.

matheusneder reviewed (commented):
@arctouch-eltonsantana 
Added wait indicator, sorry for this lapse.

Commit:
Add decoupled test for TheMovieDbService (client) with HttpMock and static mock data.

matheusneder reviewed (commented):
@arctouch-eltonsantana 
Adjusted TheMovieDbService unit test by breaking web service depency with httpmock and added the expected asserts.

matheusneder commented:
Hi @arctouch-eltonsantana 

I fixed almost issues you adressed:

* **There is no connection handling** - Please, take a look at my explanation about this https://github.com/matheusneder/upcomming-movies/pull/1#pullrequestreview-65996647
* **There is no wait indicator** - Fixed
* **Images take a long time to load** - Fixed: switched to poster images
* **Images and Http requests not cached** - Added explicity cache configuration for images, didn't cache service responses.
* **Using JObject instead of deserializing object** - Design choice
* **There is no 'Asserts' on unit tests** - Fixed

**Could you review please?**

Commit:
Minor adjustment.

matheusneder committed:
Migrate mock data to json files.
